### PR TITLE
[db] Add statusVersion to prebuilds to track status version

### PR DESCRIPTION
**Pull request by easyCZ**
## Related Issue(s)

* https://github.com/gitpod-io/gitpod/issues/8596
* https://github.com/gitpod-io/gitpod/issues/9107

## How to test

1. Preview works
2. New prebuilds can start
3. Column statusVersion appears in the DB

## Release Notes

```release-note
NONE
```

## Documentation

/hold
/uncc

### CODE REVIEW

**Comment by roboquat** [comment]
@easyCZ: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

**Review by Sunset17 — DISMISSED** [review]

File: components/gitpod-protocol/src/protocol.ts [hunk 1]
```diff
@@ -634,6 +634,7 @@ export interface PrebuiltWorkspace {
     buildWorkspaceId: string;
     creationTime: string;
     state: PrebuiltWorkspaceState;
+    statusVersion: number;
```
Comment (easyCZ) [inline components/gitpod-protocol/src/protocol.ts:637]: Not entirely happy that the statusVersion needs to be exposed in the protocol and sent to clients. Is there a way (without significantly changing the system) to have this data only available to internal components and continue not having it in clients?

Comment (geropl) [reply components/gitpod-protocol/src/protocol.ts:637]: > continue not having it in clients?

You could :
 1. introduce a sub-shape of that we use in protocol, and a more expanded shape we expose in `GitpodService`
 2. and hand-remove this field when sending it to clients (we do this with some sensitive field of `WorkspaceInstance` in `GitpodServerImpl`)
 3. but I'm not sure it warrant's the effort, now and for this single field. Let's hope make it better with the new API :wink:

**Comment by geropl** [comment]
@easyCZ I wonder if we could place the field into `WorkspaceInstance` generically, because there we'll have to have it as well. And then we have duplication. :thinking:

**Comment by easyCZ** [comment]
> @easyCZ I wonder if we could place the field into WorkspaceInstance generically, because there we'll have to have it as well. And then we have duplication. 🤔

I'd rather avoid generics for things like the DB schema and core protocol. It introduces coupling which can create unintended side-effects (especially on the DB level) when you try to break it up.

I'm happy to compose it into both, however, if you think it's needed.

File: components/gitpod-protocol/src/protocol.ts (hunk 1 above)
Comment (easyCZ) [reply components/gitpod-protocol/src/protocol.ts:637]: Thanks, that's what I was thinking as well. Unless there is a strong objection, I'll leave it as is for now.

**Comment by geropl** [comment]
> I'd rather avoid inheritance for things like the DB schema and core protocol.

Hm, I'd rather stick to the current idea of "data deduplication" (inheritance :ghost: ) even if maintain the tight coupling between both tables. If we want to split it out we have to look at it more holistically I fear.

**Comment by easyCZ** [comment]
From what I can see, the ORM definition of Prebuilds does not currently have any inheritance specified [[file](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/src/typeorm/entity/db-prebuilt-workspace.ts#L21)].

And neither does the protocol [[file]](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/src/protocol.ts#L628). Where exactly do you want to add inheritance?

I'd be open to defining an interface which has a field `statusVersion` which we can use to type the objects but YAGNI?

**Comment by geropl** [comment]
**Summary**: We're going with the statusVersion field in the prebuilds tabe
**Reasoning**:
we want to have statusVersion the API-facing shape
we agree that WorkspaceInstance will use the same field at some point
it's unpractical to read statusVersion from prebuiold (2-join Workspace->WorkspaceInstance.statusVersion)
we're taking the slight risk of data-duplication/-races in favor of
following the current pattern for API shapes (same as internal)
following the current pattern by having all fields in the shape in the DB table
Also, statusVersion tells us something about the data quality of all fields in the table (data provenance), so might make sense to duplicate it in the first place :upside_down_face:

**Review by geropl — APPROVED** [review]
LGTM

**Comment by easyCZ** [comment]
Thanks for the review and discussion @geropl!

/unhold

**Pull request merged**